ee/backend/app/src/main/kotlin/io/tolgee/ee/service/OAuthService.kt

```
@@ -0,0 +1,231 @@
+
+      val expirationTime: Date = jwtClaimsSet.expirationTime
+      if (expirationTime.before(Date())) {
+        throw OAuthAuthorizationException(Message.SSO_ID_TOKEN_EXPIRED, null)
+      }
+
+      return GenericUserResponse().apply {
+        sub = jwtClaimsSet.subject
+        name = jwtClaimsSet.getStringClaim("name")
+        given_name = jwtClaimsSet.getStringClaim("given_name")
+        family_name = jwtClaimsSet.getStringClaim("family_name")
+        email = jwtClaimsSet.getStringClaim("email")
+      }
+    } catch (e: Exception) {
+      logger.info(e.stackTraceToString())
+      throw OAuthAuthorizationException(Message.SSO_USER_INFO_RETRIEVAL_FAILED, null)
+    }
+  }
+
+  private fun register(
+    userResponse: GenericUserResponse,
+    tenant: SsoTenant,
+    invitationCode: String?,
+    refreshToken: String,
+  ): JwtAuthenticationResponse {
+    val email =
+      userResponse.email ?: let {
+        logger.info("Third party user email is null. Missing scope email?")
+        throw AuthenticationException(Message.THIRD_PARTY_AUTH_NO_EMAIL)
+      }
+    val userData =
+      OAuthUserDetails(
+        sub = userResponse.sub!!,
+        name = userResponse.name,
+        givenName = userResponse.given_name,
+        familyName = userResponse.family_name,
+        email = email,
+        domain = tenant.domain,
+        organizationId = tenant.organizationId,
+        refreshToken = refreshToken,
+      )
+    val user = oAuthUserHandler.findOrCreateUser(userData, invitationCode, ThirdPartyAuthType.SSO)
+    val jwt = jwtService.emitToken(user.id)
+    return JwtAuthenticationResponse(jwt)
+  }
+
+  override fun verifyUserIsStillEmployed(
+    ssoDomain: String?,
+    userId: Long,
+    refreshToken: String?,
+    thirdPartyAuth: String?,
+  ): Boolean {
+    val thirdPartyAuthType =
+      thirdPartyAuth?.let {
+        runCatching { ThirdPartyAuthType.valueOf(it.uppercase()) }.getOrNull()
+      }
```

We should rather fail with a bad request than fall back to `null` for invalid values. I would also do the conversion outside this function and pass the `ThirdPartyAuthType` instead.